File: app/schemas/campaign.py
"""
Pydantic schemas for campaign CRUD operations.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


class CampaignSettings(BaseModel):
    """Nested settings object for a campaign."""
    max_emails_per_day: int = 50
    delay_between_emails_seconds: int = 60
    follow_up_count: int = 3
    follow_up_delay_hours: int = 48
    tone: str = "professional"
    timezone: str = "UTC"


class CampaignCreate(BaseModel):
    """POST /api/campaigns request body."""
    name: str = Field(..., min_length=1, max_length=300)
    description: str = ""
    subject_template: str = ""
    body_template: str = ""
    gmail_account_id: str = ""
    settings: CampaignSettings = Field(default_factory=CampaignSettings)
    sequence_steps: Optional[list[dict[str, Any]]] = None


class CampaignUpdate(BaseModel):
    """PATCH /api/campaigns/{id} request body – all fields optional."""
    name: Optional[str] = None
    description: Optional[str] = None
    status: Optional[str] = None
    subject_template: Optional[str] = None
    body_template: Optional[str] = None
    gmail_account_id: Optional[str] = None
    settings: Optional[CampaignSettings] = None
    sequence_steps: Optional[list[dict[str, Any]]] = None


class CampaignResponse(BaseModel):
    """Full campaign representation returned by the API."""
    id: str
    user_id: str
    name: str
    description: str
    status: str
    subject_template: str
    body_template: str
    gmail_account_id: str
    settings: dict[str, Any]
    total_leads: int
    sent_count: int
    sequence_steps: list[dict[str, Any]] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/campaign.ts
export interface CampaignSettings {
  max_emails_per_day: number;
  delay_between_emails_seconds: number;
  follow_up_count: number;
  follow_up_delay_hours: number;
  tone: string;
  timezone: string;
}

export interface SequenceStep {
  step_number: number;
  channel: string;
  delay_days: number;
  subject_template?: string;
  body_template?: string;
  notes?: string;
}

export interface Campaign {
  id: string;
  user_id: string;
  name: string;
  description: string;
  status: string;
  subject_template: string;
  body_template: string;
  gmail_account_id: string;
  settings: CampaignSettings;
  sequence_steps?: SequenceStep[];
  total_leads: number;
  emails_sent: number;
  created_at: string;
  updated_at: string;
}

export interface CampaignCreate {
  name: string;
  description?: string;
  subject_template?: string;
  body_template?: string;
  gmail_account_id?: string;
  settings?: CampaignSettings;
  sequence_steps?: SequenceStep[];
}